pipert/contrib/database/handler.py

```diff
@@ -0,0 +1,176 @@
+from flask import Flask
+from flask_sqlalchemy import SQLAlchemy
+from sqlalchemy import exc as sql_exc, Column, Integer, ARRAY, Float, JSON, Text, DateTime
+import re
+
+
+class DBHandler:
+	def __init__(self, db_conf: dict, logger) -> None:
+		"""
+		Initializes the database handler, establishes a connection and defines the needed tables.
+
+		:param db_conf: A dictionary with *at least* the following keys: (user, pw, host, port, db) that defines the
+						connection parameters to the postgres database.
+		:param logger: logging.logger object that should be used to log any error/debug messages.
+		"""
+		self.user = None
+		self.pw = None
+		self.host = None
+		self.port = None
+		self.db_name = None
+		self.logger = logger
+		self.tables = dict()
+		self.db = None
+		self.has_connection = False
+
+		# init app and db connection
+		self.app = Flask(__name__)
+		self.app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False  # silence the deprecation warning
+		self.set_connection(**db_conf)
+
+		# init db tables
+		self._define_tables()
+
+	def insert(self, item) -> None:
+		"""
+		Attempt to insert an 'item' into the database.
+
+		'item' should be a derivative of 'self.db.Model' and should be defined (and bound to a database table) in
+		'self._define_tables(), otherwise this will fail.
+		:param item: A Derivative of 'self.db.Model'
+		:return: None
+		"""
+		if item.__class__ not in self.tables.values():
+			err_msg = f"DataBase Insertion Error: Can't insert {item}\n" \
+				f"I don't know how to insert object of type {item.__class__}"
+			self.logger.error(err_msg) if self.logger else None
+			raise TypeError(err_msg)
+
+		if self.has_connection:
+			self.db.session.add(item)
+
+			try:
+				self.db.session.commit()
+
+			# if passed data is of wrong type\shape\size
+			except sql_exc.DataError as de:
+				self.db.session.rollback()
+				self.logger.error(self._format_sqla_error(de) + f"\nError code: {de.code}")
+
+			# if passed data violates any constraints imposed by the Database, i.e. PersonAge<0, duplicate unique, etc..
+			except sql_exc.IntegrityError as ie:
+				self.db.session.rollback()
+				self.logger.error(self._format_sqla_error(ie) + f"\nError code: {ie.code}")
+
+			# if there was a connection problem
+			except sql_exc.OperationalError as oe:
+				self.db.session.rollback()
+				self.logger.critical(self._format_sqla_error(oe) + f"\nError code: {oe.code}")
+
+			# if there was any other Database related error
+			except sql_exc.SQLAlchemyError as sqlae:
+				self.db.session.rollback()
+				self.logger.error(self._format_sqla_error(sqlae) + f"\nError code: {sqlae.code}")
+
+			# if there was an unknown error
+			except Exception:
+				self.db.session.close()
+				raise
+
+	def _define_tables(self) -> None:
```

Comment:
I think that this method needs to abstract and every project will customize his tables.